From 9e2cdcf119047fb4160200b720d96c875bdffe9b Mon Sep 17 00:00:00 2001
From: Vinayakumar B <vinayakumarb@apache.org>
Date: Wed, 20 Jan 2016 21:04:21 +0530
Subject: [PATCH 1198/2748] HDFS-9661. Deadlock in DN.FsDatasetImpl between
 moveBlockAcrossStorage and createRbw (Contributed
 by ade)

(cherry picked from commit 14255786908f991fd2022480fe5575533a3dc7ce)
(cherry picked from commit eb3dcc07396616cf760088334b67fd06bf27a49a)

Change-Id: I4f378098f11b5306d18dbc584bfe532bbcba2340
---
 .../datanode/fsdataset/impl/FsDatasetImpl.java     |   11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/FsDatasetImpl.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/FsDatasetImpl.java
index 811f31c..3fc8779 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/FsDatasetImpl.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/FsDatasetImpl.java
@@ -893,8 +893,11 @@ public ReplicaInfo moveBlockAcrossStorage(ExtendedBlock block,
           + replicaInfo.getVolume().getStorageType());
     }
 
-    try (FsVolumeReference volumeRef = volumes.getNextVolume(
-        targetStorageType, block.getNumBytes())) {
+    FsVolumeReference volumeRef = null;
+    synchronized (this) {
+      volumeRef = volumes.getNextVolume(targetStorageType, block.getNumBytes());
+    }
+    try {
       File oldBlockFile = replicaInfo.getBlockFile();
       File oldMetaFile = replicaInfo.getMetaFile();
       FsVolumeImpl targetVolume = (FsVolumeImpl) volumeRef.getVolume();
@@ -913,6 +916,10 @@ public ReplicaInfo moveBlockAcrossStorage(ExtendedBlock block,
 
       removeOldReplica(replicaInfo, newReplicaInfo, oldBlockFile, oldMetaFile,
           oldBlockFile.length(), oldMetaFile.length(), block.getBlockPoolId());
+    } finally {
+      if (volumeRef != null) {
+        volumeRef.close();
+      }
     }
 
     // Replace the old block if any to reschedule the scanning.
-- 
1.7.9.5

